### PR TITLE
Move Payment account creation and removal from UI to core

### PR DESCRIPTION
**Pull request by blabno**

### CODE REVIEW

File: core/src/main/java/bisq/core/exceptions/NotFoundException.java
Comment (chimp1984) [inline core/src/main/java/bisq/core/exceptions/NotFoundException.java:20]: I think the class name should be more specific. What is not found?

File: core/src/main/java/bisq/core/payment/validation/AltCoinAddressValidator.java
Comment (chimp1984) [inline core/src/main/java/bisq/core/payment/validation/AltCoinAddressValidator.java]: If that was removed for testing consider to use the mockito-extensions like in: https://github.com/bisq-network/bisq/pull/3587/commits/6d983dec8d37a038a6e9f1cc024f941050632140#diff-1bf51dc85c38e8dd66e85a3d5286d1e0

File: core/src/main/java/bisq/core/user/Preferences.java
Comment (chimp1984) [inline core/src/main/java/bisq/core/user/Preferences.java]: If 'final' was removed for testing consider to use the mockito-extensions like in: https://github.com/bisq-network/bisq/pull/3587/commits/6d983dec8d37a038a6e9f1cc024f941050632140#diff-1bf51dc85c38e8dd66e85a3d5286d1e0

File: core/src/main/java/bisq/core/payment/PaymentAccountManager.java
Comment (chimp1984) [inline core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: I prefer to use exceptions only in exceptional cases not in valid use case (user tries to remove account but app does not allow it as its used in trades or offers). Maybe best to refactor the check out to 'canRemove()' method and only call the method if it is allowed to remove.

File: desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java
Comment (chimp1984) [inline desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java]: Why not pass the paymentAccount and avoid the extra lookup and null check in paymentAccountManager.removePaymentAccount?

File: desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java
Comment (chimp1984) [inline desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java:97]: I think we should change the API here to not return a boolean but to add a 'canRemove()' method at the caller and only call remove if that returns true.

File: core/src/main/java/bisq/core/exceptions/NotFoundException.java
Comment (blabno) [reply core/src/main/java/bisq/core/exceptions/NotFoundException.java:20]: This is general purpose exception that can be applied to any missing entity. The caller should know what is the main entity referenced by method.

File: core/src/main/java/bisq/core/payment/PaymentAccountManager.java
Comment (blabno) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: No! The method must not make any assumptions that it is being called in happy path scenario.
Exceptions are exactly for this type of situations where something else needs to be signaled then the happy path result.

File: desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java
Comment (blabno) [reply desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java]: Because this method is going to be reused also by the API. I can create 2 methods, one working with PaymentAccount and the other with working with id and delegating to the first one after the lookup.
However I thought it's better to have less code, especially that the performance penalty is insignificant.

File: desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java
Comment (blabno) [reply desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java:97]: I think it's better to propagate the exception and handle it on the view layer, but wanted to do as little changes as possible in order not to overwhelm maintainers.

File: core/src/main/java/bisq/core/payment/PaymentAccountManager.java
Comment (chimp1984) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: I disagree. I think the method removePaymentAccount should do what its name suggests and not throw an unchecked exception if the valid case happens that the user tries to remove an account but has an offer open using it. 
Bisq is using exceptions for exceptional cases not for flow control. I think we should stick with that patter and not introduce tons of exceptions for all kind of checks.

File: desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java
Comment (chimp1984) [reply desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java]: I prefer to have then 2 methods if you need that or just do the lookup in your client code.

Comment (chimp1984) [reply desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java]: We do not use ids as params for lookup if the object is available anyway.

File: core/src/main/java/bisq/core/exceptions/ValidationException.java
Comment (lusarz) [inline core/src/main/java/bisq/core/exceptions/ValidationException.java:28]: Only first one constructor with `String message` as param is used anywhere in application.

File: core/src/main/java/bisq/core/payment/PaymentAccountInUseException.java
Comment (lusarz) [inline core/src/main/java/bisq/core/payment/PaymentAccountInUseException.java]: I wonder if it's appropriate place for this Exception class. What about `bisq.core.payment.exceptions` ? 

Although I don't see consistency in project codebase - some exceptions are [inside](https://github.com/blabno/exchange/blob/2ad86abe9fc3ea802c95d7c25d1195eacb587d60/core/src/main/java/bisq/core/dao/exceptions/DaoDisabledException.java) `exceptions` package, others [no](https://github.com/blabno/exchange/blob/d55114e019fda469266dab50f63916289b2e71cb/core/src/main/java/bisq/core/support/dispute/DisputeAlreadyOpenException.java).

File: core/src/main/java/bisq/core/payment/PaymentAccountManager.java
Comment (lusarz) [inline core/src/main/java/bisq/core/payment/PaymentAccountManager.java:70]: Shouldn't `cryptoCurrencyAccount.getSingleTradeCurrency()` throws exception instead returning `null` ?

I'm just curious your opinion - such a change would require changes in many places, and it's out of scope of current pull request.

File: core/src/main/java/bisq/core/payment/PaymentAccountManager.java
Comment (lusarz) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: I see this exception is handled [here](https://github.com/blabno/exchange/blob/a98cf1b9e84ceca658692547a92b0ec45653460b/desktop/src/main/java/bisq/desktop/main/account/content/fiataccounts/FiatAccountsDataModel.java#L100-L107). Maybe `removePaymentAccount` method can return boolean directly ?

File: core/src/main/java/bisq/core/payment/PaymentAccountManager.java
Comment (lusarz) [inline core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: Is it out of scope of current pull request ?

File: desktop/src/main/java/bisq/desktop/main/account/content/fiataccounts/FiatAccountsDataModel.java
Comment (lusarz) [inline desktop/src/main/java/bisq/desktop/main/account/content/fiataccounts/FiatAccountsDataModel.java:57]: Seems like `FiatAccountsDataModel` and `AltCoinAccountsDataModel` are very similar, especially after changes in this pull request. Shouldn't we introduce some common abstraction for them ?

**Review by lusarz — CHANGES_REQUESTED** [review]

File: core/src/main/java/bisq/core/payment/PaymentAccountManager.java
Comment (chimp1984) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java:70]: Policy in Bisq is to use exceptions for exceptional cases. If a field can be null we use null checks or `Optional` but we do not throw an exception. Only if a fields is expected to be not null we use `checkNotNull` which throws an exception. I know there are different philosophies how to deal with exceptions but I think it is not a good idea to start to mix 2 different philosophies when the projects is using already one.

File: core/src/main/java/bisq/core/payment/PaymentAccountManager.java
Comment (chimp1984) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: > Maybe removePaymentAccount method can return boolean directly ?

That would be consistent with classical `remove` methods which return a boolean as well about the success of the intended remove. In that case the client (e.g. UI) need to know why it was not possible. For that I recommend using custom methods.

File: desktop/src/main/java/bisq/desktop/main/account/content/fiataccounts/FiatAccountsDataModel.java
Comment (chimp1984) [reply desktop/src/main/java/bisq/desktop/main/account/content/fiataccounts/FiatAccountsDataModel.java:57]: Yes, might be a refactor candidate if fuctionality has duplications there.

**Review by ripcurlx — CHANGES_REQUESTED** [review]
NACK - could you please address the comments by @lusarz and @chimp1984 and resolve conflicts with master. Thanks!

File: desktop/src/main/java/bisq/desktop/main/account/content/fiataccounts/FiatAccountsDataModel.java
Comment (blabno) [reply desktop/src/main/java/bisq/desktop/main/account/content/fiataccounts/FiatAccountsDataModel.java:57]: Yes, but let's keep this PR focused and small.

File: core/src/main/java/bisq/core/payment/PaymentAccountManager.java
Comment (blabno) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: Not using exceptions for flow control? What should they be used for then?

File: desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java
Comment (blabno) [reply desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java]: >We do not use ids as params for lookup if the object is available anyway.

That's cool for desktop app, but as soon as you get any kind of API you will have to reference stuff by id.

Comment (chimp1984) [reply desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java]: Then use 2 methods with diff. params. Forcing other clients to go down to id just to get back the object what is already available is not good.

File: core/src/main/java/bisq/core/payment/PaymentAccountManager.java
Comment (chimp1984) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: > Not using exceptions for flow control? What should they be used for then?

This is considered bad practice:
https://duckduckgo.com/?q=java+exceptions+for+flow+control&t=ffab&ia=web

Comment (chimp1984) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: > What should they be used for then?

For cases which happens exceptionally and sometimes lack of a clear way to handle it, so you delegate it to higher level classes how to deal with it. Writing to disk but get a IOError for instance. This use case is a valid circumstance that a user want to remove an account but does not know that there is a policy that he cannot remove it if used in offers or trades. This is not exceptional. I guess you want to provide information for the caller about the reason why the remove was not successful in the exception. You could also create a resultObject which contains diff. reuslt types and meta data if needed. Or just let the client call some validation methods like `isAccountUsedInOpenOffers` which is the usual way how Bisq handles such cases.

Comment (blabno) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: So instead of of throwing Exception we should return some Object that either contains happy path result, or some additional data in case of failure?
Exceptions are much cleaner.

File: desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java
Comment (blabno) [reply desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java]: Ok. Though I will get another comment from @lusarz that method is not used.

File: core/src/main/java/bisq/core/payment/PaymentAccountManager.java
Comment (chimp1984) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: I give up.... 
NACK to the PR

**Review by chimp1984 — CHANGES_REQUESTED** [review]
NACK, see comments. Bisq does not use exceptions for flow control which is considered by many developers as anti-pattern. Even if you tolerate that view it is bad to mix 2 different "philosophies" in a project.

File: core/src/main/java/bisq/core/payment/PaymentAccountManager.java
Comment (blabno) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: Yes, give up, cause you don't have better arguments. When there are many reasons why a method might fail then the typed exceptions are the way to inform other component about the reason.
Strings are good for humans, but it's not only the humans that interact with this business logic component.

Comment (lusarz) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: Why strings? You can create some kind of OperationResult class, like [ValidationResult](https://github.com/bisq-network/bisq/blob/049655b6857449aaca789c9cee5358daf0a54419/core/src/main/java/bisq/core/util/validation/InputValidator.java#L37-L64).

You'll be able to map it to appropriate response in API.

Comment (blabno) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: @lusarz do you honestly use OperationResult for every non-void method that you write?

**Comment by blabno** [comment]
@ripcurlx comments applied although I don't agree with some of them at all.

File: core/src/main/java/bisq/core/payment/PaymentAccountManager.java
Comment (lusarz) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: No, and I don't have strong opinion what is better - Exceptions or ResultObject. I usually use Exceptions, but now I see a lot of opinions in net that recommends ResultObject. I think this approach might be interesting. Client code with a lot of `catch` blocks doesn't looks too nice. 

Just give it a try, maybe you'll find some pros :)

Also consistency argument is relevant here. Validation in bisq [(for example here)](https://github.com/bisq-network/bisq/blob/049655b6857449aaca789c9cee5358daf0a54419/core/src/main/java/bisq/core/util/validation/InputValidator.java#L26-L28) uses ResultObject instead of throwing typed exceptions.

Comment (blabno) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: Validation is completely different case. When you call `validate` method the main thing you want to know is the result of the validation.
Look at classic example of ArrayList: http://www.docjar.com/html/api/java/util/ArrayList.java.html#441

Does remove method return RemoveMethodResult? No it returns the removed object, which is the main intended thing. In order to tell user that object cannot be removed, because index is wrong it uses IndexOutOfBoundsException.

Comment (lusarz) [reply core/src/main/java/bisq/core/payment/PaymentAccountManager.java]: Look at ArrayList [30 lines lower](http://www.docjar.com/html/api/java/util/ArrayList.java.html#472), it has `remove(Object o)` method. Does it throw `NotFoundException` ? No, it returns false.

**Comment by chimp1984** [comment]
I don't have time for another review. I leave review to other devs and if they ACK my NACK can be ignored.

**Pull request closed without merging**
